jgrapht-core/src/main/java/org/jgrapht/alg/flow/GusfieldGomoryHuCutTree.java

```diff
@@ -0,0 +1,405 @@
+     */
+    public GusfieldGomoryHuCutTree(Graph<V, E> network, double epsilon)
+    {
+        this(network, new PushRelabelMFImpl<>(network, epsilon));
+    }
+
+    /**
+     * Constructs a new GusfieldEquivalentFlowTree instance.
+     * 
+     * @param network input graph
+     * @param minimumSTCutAlgorithm algorithm used to compute the minimum s-t cuts
+     */
+    public GusfieldGomoryHuCutTree(
+        Graph<V, E> network, MinimumSTCutAlgorithm<V, E> minimumSTCutAlgorithm)
+    {
+        if(!(network instanceof UndirectedGraph))
+            throw new IllegalArgumentException("Graph must be undirected");
+        this.network = network;
+        this.N = network.vertexSet().size();
+        if (N < 2)
+            throw new IllegalArgumentException("Graph must have at least 2 vertices");
+        this.minimumSTCutAlgorithm = minimumSTCutAlgorithm;
+        vertexList.addAll(network.vertexSet());
+        for (int i = 0; i < vertexList.size(); i++)
+            indexMap.put(vertexList.get(i), i);
+    }
+
+    /**
+     * Runs the algorithm
+     */
+    private void calculateGomoryHuTree()
+    {
+        flowMatrix = new double[N][N];
+        p = new int[N];
+        fl = new double[N];
+
+        for (int s = 1; s < N; s++) {
+            int t = p[s];
+            double flowValue =
+                minimumSTCutAlgorithm.calculateMinCut(vertexList.get(s), vertexList.get(t));
+            Set<V> sourcePartition = minimumSTCutAlgorithm.getSourcePartition(); // Set X in the
+                                                                                 // paper
+            fl[s] = flowValue;
+
+            for (int i = 0; i < N; i++)
+                if (i != s && sourcePartition.contains(vertexList.get(i)) && p[i] == t)
+                    p[i] = s;
+            if (sourcePartition.contains(vertexList.get(p[t]))) {
+                p[s] = p[t];
+                p[t] = s;
+                fl[s] = fl[t];
+                fl[t] = flowValue;
+            }
+
+            // populate the flow matrix
+            flowMatrix[s][t] = flowMatrix[t][s] = flowValue;
+            for (int i = 0; i < s; i++)
+                if (i != t)
+                    flowMatrix[s][i] =
+                        flowMatrix[i][s] = Math.min(flowMatrix[s][t], flowMatrix[t][i]);
+        }
+    }
+
+    /**
+     * Returns the Gomory-Hu Tree as an actual tree (graph). Note that this tree is not necessary
+     * unique. The edge weights represent the flow values/cut weights. This method runs in O(n)
+     * time.
+     * 
+     * @return Gomory-Hu Tree
+     */
+    public SimpleWeightedGraph<V, DefaultWeightedEdge> getGomoryHuTree()
+    {
+        if (p == null) // Lazy invocation of the algorithm
+            this.calculateGomoryHuTree();
+
+        //Compute the tree from scratch. Since we compute a new tree, the user is free to modify this tree.
+        //Alternatively we could return a Unmodifiable view of this.gomoryHuTree
```

Comment:
I thought about this, but the most common use-case is to modify the returned gomory-hu tree.
For instance, there are several methods in both the GusfieldGomoryHuCutTree class and the PadbergRaoOddMinimumCutset which get the tree as a graph, and then add/delete edges from it. If we would return a unmodifiable view instead, the user first has to create a modifiable copy of the graph before he can use the graph is his implementation. This is rather annoying. Moreover, the getGomoryHuTree is a relatively cheap method, so rebuilding the tree from scratch is not a big problem.

Would you agree? If not, I'll rewrite the method to return an unmodifiable copy.